### PR TITLE
Update airtable plugin

### PR DESCRIPTION
**Pull request by bozhi**
Update `airtable_record_list` to support retrieving records by `view_id`
- Context: Code gen usually generates code that processes data in the order it is returned by `airtable_record_list` and the result to the user looks like this:
AirtableProfileLookup

Update `airtable_record_list` docstring to be more accurate regarding what is returned in fields
- Code gen would sometimes generate code that checks for the existence of a field in the record:
Screenshot 2024-02-12 at 10 35 41 AM

See also: https://lutra.ai/run-debug/workflows/deW0k5xWHXs/018d8b5a-0319-82f7-9855-8e79ff87e627

- Updating the docstring to let code gen know that fields aren't always returned.

### CODE REVIEW

File: airtable/plugin.py [hunk 1]
```diff
@@ -156,22 +165,29 @@ class AirtableRecord:
 
 
 def airtable_record_list(
-    base_id: AirtableBaseID,
```
Comment (vrv) [inline airtable/plugin.py:159]: I don't know much about Airtable but it feels like base_id is probably more 'primary' than table_id ?

File: airtable/plugin.py [hunk 2]
```diff
@@ -30,16 +38,15 @@ class AirtableRecordID:
 
 def airtable_parse_ids_from_url(
     url: str,
-) -> Tuple[AirtableBaseID, Optional[AirtableTableID], Optional[AirtableRecordID]]:
+) -> Tuple[AirtableBaseID, Optional[AirtableTableID], Optional[AirtableRecordID], Optional[AirtableViewID]]:
```
Comment (vrv) [inline airtable/plugin.py:41]: Did you try the idea of having a super-object that contained all the sub IDs instead of returning the tuple?

I feel like this exposes a lot of the detail to code gen to properly handle and pass all of the individual elements, and maybe it's better if the plumbing is simple and all APIs take AirtableID and our implementations resolve how to use it?

Comment (bozhi) [reply airtable/plugin.py:41]: Yeah, the first version I tried was
```
@dataclass
class AirtableDataContextID:
    base_id: str
    table_id: Optional[str]
    view_id: Optional[str]

def airtable_parse_ids_from_url(
    url: str
) -> Tuple[AirtableDataContextID, Optional[AirtableRecordID]:
    ...

def airtable_record_list(
    context_id: AirtableDataContextID,
    include_fields: Optional[set[str]] = None,
) list[AirtableRecord]:
    ...
```

@jcharum do you see any issue with the version with returning an object that contains the base_id, table_id, view_id? I think a variation of this was discuss before but we landed on the version that returns the ids separately.

Comment (jcharum) [reply airtable/plugin.py:41]: I have not done a systematic comparison, but that seems worse to me as it moves something that is type-checked into something that isn't.

File: airtable/plugin.py [hunk 3]
```diff
@@ -30,16 +38,15 @@ class AirtableRecordID:
...
     """
     Parses Airtable IDs from an Airtable web UI URL.  The minimum required URL format is
     https://airtable.com/{base_id}, with {base_id} being mandatory.  The URL may
     optionally include {table_id}, {view_id}, and {record_id} segments in this sequence,
-    but these are not required.  Any segments or query parameters beyond the specified
+    but these are not required. Any segments or query parameters beyond the specified
```
Comment (jcharum) [inline airtable/plugin.py:46]: Was this change intentional?  If not, undo.

File: airtable/plugin.py (hunk 2 above)
Comment (jcharum) [inline airtable/plugin.py:41]: Order the tuple the same as how they appear in the URL?  I think keeping the order consistent is useful (at a minimum for human-reading purposes), and the URL order seems sensible.

File: airtable/plugin.py [hunk 4]
```diff
@@ -58,10 +65,12 @@ def airtable_parse_ids_from_url(
         base_id = match.group("base_id")
         table_id = match.group("table_id")
         record_id = match.group("record_id")
+        view_id = match.group("view_id")
```
Comment (jcharum) [inline airtable/plugin.py:68]: Same comment about ordering.

File: airtable/plugin.py [hunk 5]
```diff
@@ -19,7 +19,15 @@ class AirtableTableID:
     """
     id can either be the table name or the table ID.
     """
+    id: str
+
 
+@dataclass
+class AirtableViewID:
+    '''
```
Comment (jcharum) [inline airtable/plugin.py:27]: double quotes

File: airtable/plugin.py [hunk 6]
```diff
@@ -159,19 +168,26 @@ def airtable_record_list(
...
-    will access by name.
-
-    If `include_fields` is not `None`, only the fields specified in `include_fields`
-    will be included.  If `include_fields` is `None`, all fields are included.
+    Returns the results of an Airtable `list records` API call, allowing selection of
+    specific fields to include in the returned records. Fields present in the schema
+    but not assigned a value will not be returned.
+
+    Important:
+    - If `include_fields` is specified (not `None`), only the fields listed in `include_fields`
```
Comment (jcharum) [inline airtable/plugin.py:179]: Does the "is specified" make a difference?  It seems redundant.

File: airtable/plugin.py [hunk 7]
```diff
@@ -159,19 +168,26 @@ def airtable_record_list(
...
-    Return results of an Airtable `list records` API call with the option of choosing
-    which fields to include in the returned records.  The include_fields fields are
-    required, so populate it to ensure that the returned records have the fields you
-    will access by name.
-
-    If `include_fields` is not `None`, only the fields specified in `include_fields`
-    will be included.  If `include_fields` is `None`, all fields are included.
+    Returns the results of an Airtable `list records` API call, allowing selection of
+    specific fields to include in the returned records. Fields present in the schema
+    but not assigned a value will not be returned.
```
Comment (jcharum) [inline airtable/plugin.py:176]: Does the part about the fields actually make a difference?  It seems only relevant to code gen, and the rest of the comment below addresses it.

File: airtable/plugin.py [hunk 8]
```diff
@@ -159,19 +168,26 @@ def airtable_record_list(
...
+
+    Important:
+    - If `include_fields` is specified (not `None`), only the fields listed in `include_fields`
+    will be included in the results.
+    - If `include_fields` is `None`, all fields that have assigned values will be included,
+    excluding any fields that are defined in the schema but are empty.
```
Comment (jcharum) [inline airtable/plugin.py:182]: "excluding" is confusing here, as the restrictive clause makes it have no intersection with "all fields that have assigned values" anyway.  That is, "excluding" isn't describing anything that's excluded from the defined set.

File: airtable/plugin.py (hunk 3 above)
Comment (bozhi) [reply airtable/plugin.py:46]: Was the double spaces intentional? I removed it because that was odd to me

File: airtable/plugin.py (hunk 2 above)
Comment (bozhi) [reply airtable/plugin.py]: To close the loop out:

```
so the issue is that some APIs require specific sets of IDs





[3:25](https://d8e-ai.slack.com/archives/D05ABC70YR0/p1707866748928529)
e.g., some APIs only take a baseid
[3:25](https://d8e-ai.slack.com/archives/D05ABC70YR0/p1707866756661279)
others take one with a record ID
[3:26](https://d8e-ai.slack.com/archives/D05ABC70YR0/p1707866769272949)
and if you have a superclass, the type doesn't define what it contains and whether it's relevant
[3:26](https://d8e-ai.slack.com/archives/D05ABC70YR0/p1707866776661009)
which is something code gen probably does need to know about
[3:27](https://d8e-ai.slack.com/archives/D05ABC70YR0/p1707866822332539)
one alternative would be some kind of "discriminated union" type, but code gen would need to check the type of the ID, which MAY be better than optionals
[3:27](https://d8e-ai.slack.com/archives/D05ABC70YR0/p1707866864801369)
but it's not clear, so probably not worth changing significantly right now
```

**Comment by bozhi** [comment]
@jcharum @vrv let me know if these addressed your concerns :)
